File: plugins/plugin-google-chat/rust/src/actions/list_spaces.rs
//! List spaces action for Google Chat plugin.

use crate::service::GoogleChatService;
use crate::types::{get_space_display_name, is_direct_message, GoogleChatSpace};
use serde::{Deserialize, Serialize};

/// Result from the list spaces action.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListSpacesResult {
    pub success: bool,
    pub space_count: usize,
    pub spaces: Vec<SpaceInfo>,
    pub formatted_text: String,
    pub error: Option<String>,
}

/// Space information for display.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpaceInfo {
    pub name: String,
    pub display_name: Option<String>,
    pub space_type: String,
    pub threaded: bool,
}

impl From<GoogleChatSpace> for SpaceInfo {
    fn from(space: GoogleChatSpace) -> Self {
        Self {
            name: space.name,
            display_name: space.display_name,
            space_type: space.space_type,
            threaded: space.threaded,
        }
    }
}

/// Execute the list spaces action.
pub async fn execute_list_spaces(service: &GoogleChatService) -> ListSpacesResult {
    let spaces = match service.get_spaces().await {
        Ok(s) => s,
        Err(e) => {
            return ListSpacesResult {
                success: false,
                space_count: 0,
                spaces: Vec::new(),
                formatted_text: String::new(),
                error: Some(e.to_string()),
            }
        }
    };

    let space_infos: Vec<SpaceInfo> = spaces.iter().map(|s| s.clone().into()).collect();

    // Format space list
    let formatted = if space_infos.is_empty() {
        "Not currently in any Google Chat spaces.".to_string()
    } else {
        let space_lines: Vec<String> = spaces
            .iter()
            .map(|space| {
                let name = get_space_display_name(space);
                let space_type = if is_direct_message(space) {
                    "DM"
                } else {
                    &space.space_type
                };
                let threaded = if space.threaded { " (threaded)" } else { "" };
                format!("• {} [{}]{}", name, space_type, threaded)
            })
            .collect();

        format!(
            "Currently in {} space(s):\n\n{}",
            space_infos.len(),
            space_lines.join("\n")
        )
    };

    ListSpacesResult {
        success: true,
        space_count: space_infos.len(),
        spaces: space_infos,
        formatted_text: formatted,
        error: None,
    }
}

File: plugins/plugin-google-chat/typescript/src/actions/listSpaces.ts
/**
 * List spaces action for Google Chat plugin.
 */

import type {
  Action,
  ActionResult,
  HandlerCallback,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import type { GoogleChatService } from "../service.js";
import type { GoogleChatSpace } from "../types.js";
import {
  GOOGLE_CHAT_SERVICE_NAME,
  getSpaceDisplayName,
  isDirectMessage,
} from "../types.js";

export const listSpaces: Action = {
  name: "GOOGLE_CHAT_LIST_SPACES",
  similes: [
    "LIST_GOOGLE_CHAT_SPACES",
    "GCHAT_SPACES",
    "SHOW_GOOGLE_CHAT_SPACES",
  ],
  description: "List all Google Chat spaces the bot is a member of",

  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
  ): Promise<boolean> => {
    return message.content.source === "google-chat";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | undefined,
    _options?: Record<string, unknown>,
    callback?: HandlerCallback,
  ): Promise<ActionResult> => {
    const gchatService = await runtime.getService<GoogleChatService>(
      GOOGLE_CHAT_SERVICE_NAME,
    );

    if (!gchatService || !gchatService.isConnected()) {
      if (callback) {
        callback({
          text: "Google Chat service is not available.",
          source: "google-chat",
        });
      }
      return { success: false, error: "Google Chat service not available" };
    }

    const spaces = await gchatService.getSpaces();

    if (spaces.length === 0) {
      if (callback) {
        callback({
          text: "I'm not currently in any Google Chat spaces.",
          source: message.content.source as string,
        });
      }
      return {
        success: true,
        data: { spaceCount: 0, spaces: [] },
      };
    }

    // Format space list
    const spaceLines = spaces.map((space: GoogleChatSpace) => {
      const name = getSpaceDisplayName(space);
      const type = isDirectMessage(space) ? "DM" : space.type || "SPACE";
      const threaded = space.threaded ? " (threaded)" : "";
      return `• ${name} [${type}]${threaded}`;
    });

    const responseText = `Currently in ${spaces.length} space(s):\n\n${spaceLines.join("\n")}`;

    if (callback) {
      callback({
        text: responseText,
        source: message.content.source as string,
      });
    }

    return {
      success: true,
      data: {
        spaceCount: spaces.length,
        spaces: spaces.map((s: GoogleChatSpace) => ({
          name: s.name,
          displayName: s.displayName,
          type: s.type,
          threaded: s.threaded,
        })),
      },
    };
  },

  examples: [
    [
      {
        name: "{{user1}}",
        content: { text: "What Google Chat spaces are you in?" },
      },
      {
        name: "{{agent}}",
        content: {
          text: "Let me check my Google Chat spaces.",
          actions: ["GOOGLE_CHAT_LIST_SPACES"],
        },
      },
    ],
    [
      {
        name: "{{user1}}",
        content: { text: "List all spaces" },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll list all the Google Chat spaces I'm in.",
          actions: ["GOOGLE_CHAT_LIST_SPACES"],
        },
      },
    ],
  ],
};
